src/hera/workflows/_meta_mixins.py

```diff
@@ -310,24 +311,37 @@ def _get_param_items_from_source(source: Callable) -> List[Parameter]:
     return non_default_parameters
 
 
-def _get_params_from_items(with_items: List[Any]) -> Optional[List[Parameter]]:
-    """Returns an optional list of `Parameter` from the specified list of `with_items`.
+def _get_parameters_used_in_with_items(with_items: Any) -> Optional[List[Parameter]]:
+    """Get the optional list of Parameters used in with_items, with their values set to corresponding templated item strings.
 
     The assembled list of `Parameter` contains all the unique parameters identified from the `with_items` list. For
     example, if the `with_items` list contains 3 serializable elements such as
     `[{'a': 1, 'b': 2}, {'a': 3, 'b': 4}, {'a': 5, 'b': 6}]`, then only 2 `Parameter`s are returned. Namely, only
-     `Parameter(name='a')` and `Parameter(name='b')` is returned, with values `{{item.a}}` and `{{item.b}}`,
-     respectively. This helps with the parallel/serial processing of the supplied items.
+    `Parameter(name='a')` and `Parameter(name='b')` is returned, with values `{{item.a}}` and `{{item.b}}`,
+    respectively. This helps with the parallel/serial processing of the supplied items. If the list is of
+    plain values and not key-values, then `None` is returned as there are no parameter substitutions.
     """
-    if len(with_items) == 0:
+    if not with_items:
         return None
-    elif len(with_items) == 1:
-        el = with_items[0]
-        if len(el.keys()) == 1:
-            return [Parameter(name=n, value="{{item}}") for n in el.keys()]
-        else:
-            return [Parameter(name=n, value=f"{{{{item.{n}}}}}") for n in el.keys()]
-    return [Parameter(name=n, value=f"{{{{item.{n}}}}}") for n in with_items[0].keys()]
+
+    if isinstance(with_items[0], dict):
```

Comment:
Worth validating that all items are *not* dicts in an else path? in case someone passes in a mixture of plain values and dicts?

Reply:
I originally had a check for "all types in the list are the same", but I'm not sure it's necessary as they'll get serialised anyway (or the user should use `shared.serialization.serialize` as in the error messages). But I see what you're saying is to make sure there aren't any dicts in the list - I do think a check for all types being the same would be a helpful thing though, wdyt? Otherwise yes, happy to check no dicts in the list 🚀

Reply:
I think checking for dicts vs not-dicts specifically is helpful because it would result in a broken outcome to do that. I don't know if it's necessary to check that all types are the same.